azure/services/privatedns/privatedns.go

```diff
@@ -19,205 +19,253 @@ package privatedns
-			err = s.client.DeleteLink(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName, linkSpec.LinkName)
-			if err != nil && !azure.ResourceNotFound(err) {
-				return errors.Wrapf(err, "failed to delete virtual network link %s with zone %s in resource group %s", linkSpec.VNetName, zoneSpec.ZoneName, s.Scope.ResourceGroup())
-			}
+			return errors.Wrapf(err, "could not get vnet link state of %s in resource group %s",
+				linkSpec.OwnerResourceName(), linkSpec.ResourceGroupName())
 		}
-		// Skip the deletion of private DNS zone which is not managed by capz.
-		isManaged, err := s.isPrivateDNSManaged(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName)
-		if err != nil && !azure.ResourceNotFound(err) {
-			return errors.Wrapf(err, "could not get private DNS zone state of %s in resource group %s", zoneSpec.ZoneName, s.Scope.ResourceGroup())
+
+		if !isVnetLinkManaged {
+			log.V(2).Info("Skipping vnet link deletion for unmanaged vnet link", "vnet link",
+				linkSpec.ResourceName(), "private dns zone", linkSpec.OwnerResourceName())
+			continue
 		}
-		if !isManaged {
-			log.V(1).Info("Skipping private DNS zone deletion for unmanaged private DNS zone", "private DNS", zoneSpec.ZoneName)
-			return nil
+
+		if err := s.vnetLinkReconciler.DeleteResource(ctx, linkSpec, serviceName); err != nil {
+			if !azure.IsOperationNotDoneError(err) || resErr == nil {
+				resErr = err
+			}
 		}
-		// Delete the private DNS zone, which also deletes all records.
-		log.V(2).Info("deleting private dns zone", "private dns zone", zoneSpec.ZoneName)
-		err = s.client.DeleteZone(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName)
-		if err != nil && azure.ResourceNotFound(err) {
-			// already deleted
+	}
+
+	return resErr
+}
+
+func (s *Service) deleteZone(ctx context.Context, zoneSpec azure.ResourceSpecGetter) error {
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "privatedns.Service.deleteZone")
+	defer done()
+
+	// Skip the reconciliation of private DNS zone which is not managed by capz.
+	isManaged, err := s.isPrivateDNSManaged(ctx, zoneSpec)
+	if err != nil {
+		if azure.ResourceNotFound(err) {
+			// already deleted or doesn't exist, cleanup status and return.
+			s.Scope.DeleteLongRunningOperationState(zoneSpec.ResourceName(), serviceName)
 			return nil
 		}
-		if err != nil && !azure.ResourceNotFound(err) {
-			return errors.Wrapf(err, "failed to delete private dns zone %s in resource group %s", zoneSpec.ZoneName, s.Scope.ResourceGroup())
-		}
-		log.V(2).Info("successfully deleted private dns zone", "private dns zone", zoneSpec.ZoneName)
+		return errors.Wrapf(err, "could not get private DNS zone state of %s in resource group %s", zoneSpec.ResourceName(), zoneSpec.ResourceGroupName())
 	}
-	return nil
+
+	if !isManaged {
+		log.V(1).Info("Skipping deletion of unmanaged private DNS zone", "private DNS", zoneSpec.ResourceName())
+		return nil
+	}
+
+	err = s.zoneReconciler.DeleteResource(ctx, zoneSpec, serviceName)
```

Comment:
Maybe add this comment `// Delete the private DNS zone, which also deletes all records.` from the old reconciler here.

Reply:
Good catch, added it.